src/rgw/rgw_rest_s3.h

```diff
@@ -862,39 +862,41 @@ inline int valid_s3_bucket_name(const std::string& name, bool relaxed=false)
     if (!relaxed)
       return -ERR_INVALID_BUCKET_NAME;
 
-  for (const char *s = name.c_str(); *s; ++s) {
-    char c = *s;
-    if (isdigit(c))
+  for (auto itr = name.begin(); itr != name.end(); itr++) {
+    char c = *itr;
+
+    if (isdigit(c)) {
       continue;
+    }
 
-    if (isalpha(c)) {
-      // name cannot contain uppercase letters
-      if (relaxed || islower(c))
-	continue;
+    if (isalpha(c) && (relaxed || islower(c))) {
+      // name cannot contain uppercase letters except in relaxed mode
+      continue;
     }
 
-    if (c == '_')
-      // name cannot contain underscore
-      if (relaxed)
-	continue;
+    if (c == '_' && relaxed) {
+      // name cannot contain underscore except in relaxed mode
+      continue;
+    }
 
-    if (c == '-')
+    if (c == '-') {
       continue;
+    }
 
-    if (c == '.') {
-      if (!relaxed && s && *s) {
-	// name cannot have consecutive periods or dashes
-	// adjacent to periods
-	// ensure s is neither the first nor the last character
-	char p = *(s-1);
-	char n = *(s+1);
-	if ((p != '-') && (n != '.') && (n != '-'))
-	  continue;
-      } else {
-	continue;
+    if(c == '.') {
+      if(relaxed) {
+        continue;
       }
-    }
 
+      // name cannot contain a dot or a dash adjacent to a dot
+      if(
+        (itr == name.begin() || (*(itr-1) != '.' && *(itr-1) != '-')) &&
```

Comment:
in the original code we check that `p != '-'` we do not check for not being a dot